### PR TITLE
feat: improve `BlockToolAdapter`

### PR DESCRIPTION
**Pull request by himself65**
## Checklist
- [x] Native inputs handling
- Input element
- Textarea element
- Handle non-text input elements
- [x] Contenteditable element handling
- [x] improve playground

- supported `backspace` and `delete` for native inputs
Left todo for native inputs (support all deletion cases manually)

### CODE REVIEW

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 1]
```diff
@@ -171,39 +207,70 @@ export class BlockToolAdapter {
...
 
-    const action = event.detail.action;
+    if (isNativeInput) {
+      const currentElement = input as HTMLInputElement | HTMLTextAreaElement;
+      const start = currentElement.selectionStart!;
+      const end = currentElement.selectionEnd!;
+
+      const action = event.detail.action;
+
+      switch (action) {
```
Comment (himself65) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:217]: I'm a little worried about this part of logic that doesn't use the `caretAdapter` class. Because onUpdate callback cannot get any information about the update from caretEvent

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 2]
```diff
@@ -73,74 +82,99 @@ export class BlockToolAdapter {
    * @param input - input element
    * @param key - data key input is attached to
    */
-  #handleBeforeInputEvent(event: InputEvent, input: HTMLElement, key: DataKey): void {
+  #handleBeforeInputEvent = (event: InputEvent, input: HTMLElement, key: DataKey): void => {
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:85]: If you want to save context here, would be better to use an anonymous arrow function when adding a listener. Arrow function methods are treated as properties, therefore created for each instance (which we expect a lot for a document)

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 3]
```diff
@@ -73,74 +82,99 @@ export class BlockToolAdapter {
...
     event.preventDefault();
 
+    const isNativeInput = this.#mode === InputMode.Native;
     const inputType = event.inputType as InputType;
-    const targetRanges = event.getTargetRanges();
-    const range = targetRanges[0];
 
-    const start = getAbsoluteRangeOffset(input, range.startContainer, range.startOffset);
-    const end = getAbsoluteRangeOffset(input, range.endContainer, range.endOffset);
+    if (isNativeInput) {
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:94]: Arguably, would be more maintainable to use the condition in each relevant case rather than creating two switch statements

File: packages/dom-adapters/src/BlockToolAdapter/index.ts (hunk 1 above)
Comment (gohabereg) [reply packages/dom-adapters/src/BlockToolAdapter/index.ts:217]: Agree, would be better if CaretAdapter handles native inputs selection on its own. I guess you can leave a to-do here to consider that for caret adapter re-implementation

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 4]
```diff
@@ -171,39 +207,70 @@ export class BlockToolAdapter {
...
+          currentElement.value = currentElement.value.slice(0, start) + text + currentElement.value.slice(end);
 
-    const start = rangeIndex[0];
-    const end = rangeIndex[1];
+          currentElement.setSelectionRange(start + text.length, start + text.length);
+          break;
+        }
+        case EventAction.Removed: {
+          if (start === end) {
+            currentElement.value = currentElement.value.slice(0, start - 1) +
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:228]: Current value could be assigned to a variable before switch statement and reused across the method

File: packages/dom-adapters/src/utils/index.ts [hunk 5]
```diff
@@ -1,2 +1,14 @@
 export * from './getAbsoluteRangeOffset.js';
 export * from './getRelativeIndex.js';
+
+export enum NativeInput {
+  Textarea = 'TEXTAREA',
+  Input = 'INPUT',
```
Comment (gohabereg) [inline packages/dom-adapters/src/utils/index.ts:6]: There are some non-text input types, but I think we shouldn't worry about them. Just throw an error if a tool tries to attach such input to adapter. Those inputs should be treated as value nodes and handled on a tool side.

What do you think @neSpecc @ilyamore88 ?

Comment (neSpecc) [reply packages/dom-adapters/src/utils/index.ts:6]: Yes. But it should be handled explicitly. Here you can check the input types supporting `selectionStart`: https://html.spec.whatwg.org/multipage/input.html#concept-input-apply:~:text=·-,selectionStart,-·

File: packages/dom-adapters/src/utils/index.ts
Comment (neSpecc) [inline packages/dom-adapters/src/utils/index.ts:1]: I'd suggest to leave `index.ts` just as index file and move your utils to `utils/input.ts`

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 6]
```diff
@@ -29,6 +32,11 @@ export class BlockToolAdapter {
    */
   #blockIndex: number;
 
+  /**
+   * Input mode
+   */
+  #mode: InputMode = InputMode.Native;
```
Comment (neSpecc) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:38]: Block Tool can have several inputs of different type

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 7]
```diff
@@ -50,10 +58,11 @@ export class BlockToolAdapter {
   public attachInput(key: DataKey, input: HTMLElement): void {
     const inputTag = input.tagName as NativeInput;
 
-    /**
-     * @todo Filter non-text-editable inputs
-     */
-    if (![NativeInput.Textarea, NativeInput.Input].includes(inputTag) && !input.isContentEditable) {
+    if (Boolean(NATIVE_INPUT_SET.has(inputTag))) {
```
Comment (neSpecc) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:61]: maybe create `isNativeInput()` util instead of this set for better readability?

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 8]
```diff
@@ -133,14 +156,17 @@ export class BlockToolAdapter {
       case InputType.DeleteSoftLineForward:
       case InputType.DeleteWordBackward:
       case InputType.DeleteWordForward: {
+        if (isNativeInput && start > 0) {
+          start = start - 1;
+        }
```
Comment (neSpecc) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:161]: add a comment here please

File: packages/playground/src/components/Input.vue [hunk 9]
```diff
@@ -20,6 +20,16 @@ const props = defineProps<{
   model: EditorJSModel;
 }>();
 
+// three types of inputs:
+// ?type=contenteditable
+// ?type=input
+// ?type=textarea
+let type = (new URL(window.location.href).searchParams.get('type') ?? 'contenteditable') as 'contenteditable' | 'input' | 'textarea';
```
Comment (neSpecc) [inline packages/playground/src/components/Input.vue:27]: you can use vue component props instead

Comment (himself65) [reply packages/playground/src/components/Input.vue:27]: I think it will be simpler if I read it directly.

**Review by himself65 — COMMENTED** [review]
for html textarea, should we create a new paragraph or something when user press `Enter`?

**Comment by gohabereg** [comment]
> for html textarea, should we create a new paragraph or something when user press `Enter`?

No, we'd like to preserve browser behaviour there. If some other behaviour is desired by a tool it should be implemented by the tool

**Comment by himself65** [comment]
> We'd like to preserve browser behavior there.

Is it a good idea that keep the default behavior?

image

Benefit is we don't need to handle all cases for cross-browser manually, but I think there would be some issue when bringing the collaborative feature in

File: packages/playground/src/components/Input.vue (hunk 9 above)
Comment (neSpecc) [reply packages/playground/src/components/Input.vue:27]: I still think that props are more suitable here. Query params are related to the page itself, when props are related to the component instance. Maybe we'll need several Inputs of different types on the Playground.

Props will be quite simpler:

**Input.vue**

```ts
defineProps<{
  type: 'contenteditable' | 'input' | 'textarea'
}>()
```

**Playground.vue**

```html
<Input type="contenteditable" />
<Input type="input" />
```

Comment (himself65) [reply packages/playground/src/components/Input.vue:27]: For now, we don't support multiple input elements on the same page, and cannot auto-unmount input elements.

File: packages/dom-adapters/package.json [hunk 10]
```diff
@@ -6,8 +6,8 @@
   "types": "dist/index.d.ts",
   "type": "module",
   "scripts": {
-    "build": "tsc --build tsconfig.build.json",
-    "dev": "yarn build --watch",
+    "build": "tsc --project tsconfig.build.json",
+    "dev": "tsc --project tsconfig.build.json --watch",
```
Comment (neSpecc) [inline packages/dom-adapters/package.json:10]: Please, describe these changes in a PR description

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 11]
```diff
@@ -133,14 +153,24 @@ export class BlockToolAdapter {
...
+        if (nativeInput && start > 0 && start === end) {
+          // for native input elements, we need to handle backspace manually
+          start = start - 1;
+        }
+        switch (inputType) {
+          case InputType.DeleteSoftLineBackward: {
+            start = 0;
+            break;
+          }
+        }
```
Comment (neSpecc) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:165]: this switch seems redundant since you're already in this case

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 12]
```diff
@@ -133,14 +153,24 @@ export class BlockToolAdapter {
       case InputType.DeleteSoftLineForward:
       case InputType.DeleteWordBackward:
       case InputType.DeleteWordForward: {
+        if (nativeInput && start > 0 && start === end) {
+          // for native input elements, we need to handle backspace manually
+          start = start - 1;
+        }
```
Comment (neSpecc) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:159]: I'd suggest not to mix different implementation in a single code. Lets move them to separate methods, like 

`handleDeleteInNativeInput()`
`handleDeleteInContentEditable()`

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 13]
```diff
@@ -171,39 +203,83 @@ export class BlockToolAdapter {
       return;
     }
 
-    const action = event.detail.action;
+    if (nativeInput) {
+      const currentElement = input as HTMLInputElement | HTMLTextAreaElement;
```
Comment (neSpecc) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:207]: same suggestion, let's move logic to separate methods

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 14]
```diff
@@ -171,39 +203,83 @@ export class BlockToolAdapter {
...
+      const [ [start, end] ] = event.detail.index;
+
+      // todo:
+      //  caretAdapter cannot handle native input
+      //  because it doesn't have the content of the input
```
Comment (neSpecc) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:212]: Please, use block-style comments and JSDoc-style todos as we use in a remaining code
```suggestion
      /** 
       * @todo caretAdapter cannot handle native because it doesn't have the content of the input
       */
```

File: packages/dom-adapters/src/caret/utils/useSelectionChange.ts [hunk 15]
```diff
@@ -52,6 +52,11 @@ export const useSelectionChange = createSingleton(() => {
       return false;
     }
 
+    if (selection.rangeCount <= 0) {
+      // special case for native input elements
+      return false;
+    }
```
Comment (neSpecc) [inline packages/dom-adapters/src/caret/utils/useSelectionChange.ts:58]: 1) Please use block-style comment
2) Put it above to related code 
3) Explain more details: why it was added, which case you're mentioning

File: packages/dom-adapters/src/caret/CaretAdapter.ts [hunk 16]
```diff
@@ -187,17 +191,24 @@ export class CaretAdapter extends EventTarget {
...
-    const end = getBoundaryPointByAbsoluteOffset(this.#input!, textIndex[1]);
+    const input: HTMLElement = this.#input!;
 
-    const selection = document.getSelection()!;
-    const range = new Range();
+    if (isNativeInput(input)) {
+      // skip
+    } else {
+      const start = getBoundaryPointByAbsoluteOffset(input,
+        textIndex[0]);
```
Comment (neSpecc) [inline packages/dom-adapters/src/caret/CaretAdapter.ts:200]: bad linebreak

File: packages/dom-adapters/src/caret/CaretAdapter.ts [hunk 17]
```diff
@@ -187,17 +191,24 @@ export class CaretAdapter extends EventTarget {
...
 
-    const start = getBoundaryPointByAbsoluteOffset(this.#input!, textIndex[0]);
-    const end = getBoundaryPointByAbsoluteOffset(this.#input!, textIndex[1]);
+    const input: HTMLElement = this.#input!;
 
-    const selection = document.getSelection()!;
-    const range = new Range();
+    if (isNativeInput(input)) {
+      // skip
+    } else {
```
Comment (neSpecc) [inline packages/dom-adapters/src/caret/CaretAdapter.ts:198]:
```suggestion
    /**
     * ...
     */
    if (isNativeInput(input)) {
      return;
    } 
```

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 18]
```diff
@@ -73,18 +79,31 @@ export class BlockToolAdapter {
...
    * @param key - data key input is attached to
    */
-  #handleBeforeInputEvent(event: InputEvent, input: HTMLElement, key: DataKey): void {
+  #handleBeforeInputEvent = (event: InputEvent, input: HTMLElement, key: DataKey): void => {
     /**
      * We prevent all events to handle them manually via model update
      */
     event.preventDefault();
 
+    const nativeInput = this.#mode === InputMode.Native;
```
Comment (neSpecc) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:88]: it is the main problem right now, BlockToolAdapter should support several inputs

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 19]
```diff
@@ -48,22 +50,68 @@ export class BlockToolAdapter {
...
+    const targetRanges = event.getTargetRanges();
+    const range = targetRanges[0];
+
+    start = getAbsoluteRangeOffset(input, range.startContainer, range.startOffset);
+    end = getAbsoluteRangeOffset(input, range.endContainer, range.endOffset);
+
+    if (start > 0 && start === end) {
+      start = start - 1;
+    }
+    switch (inputType) {
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:106]: Please add some comments for the if statement, it's not clear why it is needed

Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:106]:
```suggestion
    if (start > 0 && start === end) {
      start = start - 1;
    }

    switch (inputType) {
```

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 20]
```diff
@@ -48,22 +50,68 @@ export class BlockToolAdapter {
...
+   *
+   * @param event - beforeinput event
+   * @param input - input element
+   * @param key - data key input is attached to
+   * @private
+   */
+  #handleDeleteInNativeInput = (event: InputEvent, input: HTMLInputElement | HTMLTextAreaElement, key: DataKey): void => {
+    const inputType = event.inputType as InputType;
+    let start = input.selectionStart as number;
+    const end = input.selectionEnd as number;
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:71]: If value could be null we need to handle that case to avoid runtime errors

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 21]
```diff
@@ -48,22 +50,68 @@ export class BlockToolAdapter {
...
+
+    if (start > 0 && start === end) {
+      start = start - 1;
+    }
+    switch (inputType) {
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:76]: Please add a comment why this is required
```suggestion
    if (start > 0 && start === end) {
      start = start - 1;
    }

   switch (inputType) {
```

Comment (gohabereg) [reply packages/dom-adapters/src/BlockToolAdapter/index.ts:76]: Also in case this is a forward delete, we'd need to increase `end` by one

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 22]
```diff
@@ -48,22 +50,68 @@ export class BlockToolAdapter {
...
+      case InputType.DeleteSoftLineBackward: {
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:77]: Do we need to handle forward delete the same way here?

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 23]
```diff
@@ -133,40 +195,82 @@ export class BlockToolAdapter {
...
-  /**
-   * Handles model update events and updates DOM
-   *
-   * @param event - model update event
-   * @param input - attched input element
-   * @param key - data key input is attached to
-   * @param caretAdapter - caret adapter instance
-   */
-  #handleModelUpdate(event: ModelEvents, input: HTMLElement, key: DataKey, caretAdapter: CaretAdapter): void {
+  #handleModelUpdateNative = (event: ModelEvents, input: HTMLInputElement | HTMLTextAreaElement, key: DataKey): void => {
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:210]:
```suggestion
  #handleModelUpdateForNativeInput = (event: ModelEvents, input: HTMLInputElement | HTMLTextAreaElement, key: DataKey): void => {
```

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 24]
```diff
@@ -133,40 +195,82 @@ export class BlockToolAdapter {
...
+          currentElement.value = currentElement.value.slice(0, start) +
+            currentElement.value.slice(end);
+          currentElement.setSelectionRange(start, start);
+          break;
+        }
+      }
+    }
+  };
+
+  #handleModelUpdateContentEditable = (event: ModelEvents, input: HTMLElement, key: DataKey, caretAdapter: CaretAdapter): void => {
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:263]:
```suggestion
  #handleModelUpdateForContentEditableElement = (event: ModelEvents, input: HTMLElement, key: DataKey, caretAdapter: CaretAdapter): void => {
```

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 25]
```diff
@@ -133,40 +195,82 @@ export class BlockToolAdapter {
...
-   * @param key - data key input is attached to
-   * @param caretAdapter - caret adapter instance
-   */
-  #handleModelUpdate(event: ModelEvents, input: HTMLElement, key: DataKey, caretAdapter: CaretAdapter): void {
+  #handleModelUpdateNative = (event: ModelEvents, input: HTMLInputElement | HTMLTextAreaElement, key: DataKey): void => {
+    if (!(event instanceof TextAddedEvent) && !(event instanceof TextRemovedEvent)) {
+      return;
+    }
+
+    const [, dataIndex, blockIndex] = event.detail.index;
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:215]: We use `dataKey` in other places, let's keep it for consistency

```suggestion
    const [, dataKey, blockIndex] = event.detail.index;
```

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 26]
```diff
@@ -133,40 +195,82 @@ export class BlockToolAdapter {
...
+
+    if (blockIndex !== this.#blockIndex) {
+      return;
+    }
+
+    if (dataIndex !== composeDataIndex(key)) {
+      return;
+    }
+
+    const currentElement = input as HTMLInputElement | HTMLTextAreaElement;
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:225]: Seems to be unnecessary cast here, `input` already has `HTMLInputElement | HTMLTextAreaElement` type

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 27]
```diff
@@ -133,40 +195,82 @@ export class BlockToolAdapter {
...
+        const prevValue = currentElement.value;
+
+        currentElement.value = prevValue.slice(0, start) + text + prevValue.slice(end - 1);
+
+        currentElement.setSelectionRange(start + text.length, start + text.length);
+        break;
+      }
+      case EventAction.Removed: {
+        if (
+          start === end &&
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:242]: I don't think we can get this case from the model update. If something is deleted, start would not be equal to end

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 28]
```diff
@@ -133,40 +195,82 @@ export class BlockToolAdapter {
...
+          start > 0
+        ) {
+          currentElement.value = currentElement.value.slice(0, start - 1) +
+            currentElement.value.slice(end);
+          if (start === 0) {
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:247]: In the parent statement you check if `start > 0`, so this condition will always be false

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 29]
```diff
@@ -176,7 +280,8 @@ export class BlockToolAdapter {
     const start = rangeIndex[0];
     const end = rangeIndex[1];
 
-    const [startNode, startOffset] = getBoundaryPointByAbsoluteOffset(input, start);
+    const [startNode, startOffset] = getBoundaryPointByAbsoluteOffset(input,
+      start);
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:284]: Seems to be unnecessary change here

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 30]
```diff
@@ -133,40 +195,82 @@ export class BlockToolAdapter {
...
+  };
+
+  #handleModelUpdateContentEditable = (event: ModelEvents, input: HTMLElement, key: DataKey, caretAdapter: CaretAdapter): void => {
     if (!(event instanceof TextAddedEvent) && !(event instanceof TextRemovedEvent)) {
       return;
     }
 
     const [rangeIndex, dataIndex, blockIndex] = event.detail.index;
 
-    /**
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:160]: Could you please revert the comments

File: packages/dom-adapters/src/utils/index.ts (hunk 5 above)
Comment (gohabereg) [reply packages/dom-adapters/src/utils/index.ts:6]: Still relevant

**Review by gohabereg — COMMENTED** [review]
Not covered cases:
- Forward delete of a single character in native `INPUT` element
- Forward soft line delete in native `INPUT` element
- `insertFromDrop` event type in native `INPUT` element
- Likely `hardLineDelete` event type doesn't work at all
- `deleteWordBackward/Forward` event type doesn't work for native `INPUT` element
- The same problems for native `TEXTAREA` element
- Buggy behaviour for `deleteWordForward` for contenteditable element if caret is at the end of the line
- `deleteSoftLineForward` event type is not handled for contenteditable elements
- `deleteHardLineBackward/Forward` event types are not handled for contenteditable elements
- `insertFromDrop` event type is not handled at all
- `deleteEntireSoftLine` event type is not handled at all

I think we can merge this PR first and then work on each case in a separate PR

File: packages/dom-adapters/src/utils/index.ts (hunk 5 above)
Comment (e11sy) [reply packages/dom-adapters/src/utils/index.ts:6]: image
handled

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 31]
```diff
@@ -73,18 +130,31 @@ export class BlockToolAdapter {
    * @param input - input element
    * @param key - data key input is attached to
    */
-  #handleBeforeInputEvent(event: InputEvent, input: HTMLElement, key: DataKey): void {
+  #handleBeforeInputEvent = (event: InputEvent, input: HTMLElement, key: DataKey): void => {
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:133]: Looks like this could be a regular method, not an arrow function

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 32]
```diff
@@ -73,18 +130,31 @@ export class BlockToolAdapter {
...
     /**
      * We prevent all events to handle them manually via model update
      */
     event.preventDefault();
 
+    const nativeInput = isNativeInput(input);
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:139]: Better to call such variables with `is`, as `isNativeInput` is already declared, you could name it smth like
```suggestion
    const isInputNative = isNativeInput(input);
```

File: packages/dom-adapters/src/BlockToolAdapter/index.ts [hunk 33]
```diff
@@ -205,5 +310,23 @@ export class BlockToolAdapter {
...
+  /**
+   * Handles model update events and updates DOM
+   *
+   * @param event - model update event
+   * @param input - attched input element
+   * @param key - data key input is attached to
+   * @param caretAdapter - caret adapter instance
+   */
+  #handleModelUpdate = (event: ModelEvents, input: HTMLElement, key: DataKey, caretAdapter: CaretAdapter): void => {
+    const nativeInput = isNativeInput(input);
```
Comment (gohabereg) [inline packages/dom-adapters/src/BlockToolAdapter/index.ts:324]: Same here

File: packages/dom-adapters/src/caret/utils/useSelectionChange.ts [hunk 34]
```diff
@@ -48,7 +48,14 @@ export const useSelectionChange = createSingleton(() => {
    * @param input - input to check
    */
   function isSelectionRelatedToInput(selection: Selection | null, input: InputWithCaret): boolean {
-    if (!selection) {
+    if (selection === null) {
+      return false;
+    }
+
+    /**
+     * Check, that selection range exists in native inputs
```
[the remark this answers is not in the document]
Comment (e11sy) [reply packages/dom-adapters/src/caret/utils/useSelectionChange.ts:56]: this check turned out to be redundant

**Review by gohabereg — APPROVED** [review]

File: packages/playground/src/components/Input.vue
Comment (neSpecc) [inline packages/playground/src/components/Input.vue:1]: add style `border: 0` please to remove default browser border
image

**Review by neSpecc — APPROVED** [review]
